src/abi/mod.rs

```diff
@@ -0,0 +1,70 @@
+use crate::cargo::{manifest::CargoManifestPath, metadata::CrateMetadata};
+use crate::util;
+use near_sdk::__private::{AbiMetadata, AbiRoot};
+use std::collections::HashMap;
+use std::{fs, path::PathBuf};
+
+mod generation;
+
+const ABI_FILE: &str = "abi.json";
+
+/// ABI generation result.
+pub(crate) struct AbiResult {
+    /// Path to the resulting ABI file.
+    pub path: PathBuf,
+}
+
+pub(crate) fn execute(manifest_path: &CargoManifestPath) -> anyhow::Result<AbiResult> {
+    let tmp_dir = tempfile::Builder::new()
+        .prefix("cargo-contract_")
+        .tempdir()?;
+    log::debug!(
+        "Using temp Cargo workspace at '{}'",
+        tmp_dir.path().display()
+    );
+
+    let crate_metadata = CrateMetadata::collect(manifest_path)?;
+    let dylib_path = util::compile_dylib_project(manifest_path)?;
+
+    let cargo_toml = generation::generate_toml(manifest_path)?;
+    fs::write(tmp_dir.path().join("Cargo.toml"), cargo_toml)?;
+
+    let build_rs = generation::generate_build_rs(&dylib_path)?;
+    fs::write(tmp_dir.path().join("build.rs"), build_rs)?;
+
+    let main_rs = generation::generate_main_rs(&dylib_path)?;
+    fs::write(tmp_dir.path().join("main.rs"), main_rs)?;
+
+    let target_dir_arg = format!(
+        "--target-dir={}",
+        crate_metadata.target_directory.to_string_lossy()
+    );
+    let stdout = util::invoke_cargo(
+        "run",
+        &["--package", "near-abi-gen", "--release", &target_dir_arg],
```

Comment:
This caches the build artifacts in the target folder of the contract, and not the temp directory. Is this intended?

Reply:
Yes, I assumed this way I would be able to leverage incremental compilation if needed. Is there an argument for keeping this in the tmp directory? I guess the build artifact sizes can be pretty significant.

Reply:
Yeah, that makes sense. Just often used to having source and build artifacts together. Maybe we should consider keeping the builder source together with its build artifacts instead of a temp dir?

Reply:
yeah, I would intuitively think that keeping as is and not using a tmp directory for this would be ideal. Unsure how much would be saved by the incremental compilation, but I don't see the benefit in using tmp directory here

Reply:
Done. This makes much more sense, thanks for the suggestions.

> Unsure how much would be saved by the incremental compilation

It is quite considerable since the generated `near-abi-gen` has to depend on `near-sdk` with all of its dependencies. So always losing the target directory does not seem like a good way to go about this.